### ISSUE TITLE
[HOLD for payment 2023-02-07] [$1000] Web - Log out - Console error is displayed when log out

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#14350]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to staging.new.expensify.com
2. Log in with expensifail account
3. Go to Setting - Sign out

## Expected Result:
Console error is not displayed when log out

## Actual Result:
Console error is displayed when log out

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?

- [x] Web / Chrome / Safari

**Version Number:** 1.2.55.0

**Reproducible in staging?:** Yes

**Reproducible in production?:** No

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):** applausetester+0901abb@applause.expensifail.com/Feya87Katya

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/212787248-2d3b7542-68fc-449b-8409-b4130d1745bb.mp4

![6](https://user-images.githubusercontent.com/93399543/212787264-bee8b595-b155-41a1-be0c-539be6e4f035.jpg)
![7](https://user-images.githubusercontent.com/93399543/212787266-b92c11fa-4075-4a2f-88c7-fb765547204b.jpg)
![8](https://user-images.githubusercontent.com/93399543/212787268-251959af-2c8d-4d23-9edb-398b33dd2d9e.jpg)
![9](https://user-images.githubusercontent.com/93399543/212787270-b214dfcd-65a3-4264-9c71-c8514d005d33.jpg)

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~015d869c70481d179b</li>
        <li>Upwork Job ID: 1617488772668575744</li>
        <li>Last Price Increase: 2023-01-23</li>
    </ul>
</details>

**Comment by marcaaron** [comment]
I can reproduce this. Suspect it happens because we have already disconnected from Pusher when we try to unsubscribe from the report channel here:

https://github.com/Expensify/App/blob/3661b57ffc8aab9d7bf2b5b90f45d9de45e52978/src/pages/home/report/ReportActionsView.js#L248

The only PR I found related to logging out is here: https://github.com/Expensify/App/pull/13886/files

But not really connecting that change with the Pusher error. Probably will require some debugging.

Not sure if this is a "blocker" though. The error does look a bit harmless - but obviously would be ideal to not see it at all.

I want to say one solution would be to prevent unsubscribing if there is no websocket connection, but not really sure if that's correct to assume. And also doesn't quite explain the root cause of why this error has begun showing up.

**Comment by marcaaron** [comment]
**Update:** Only managed to reproduce 1/5 sign outs on staging - so I _think_ most likely this is harmless and should be downgraded to "not blocker" while we continue to investigate - removing the label.

**Comment by tgolen** [comment]
Yeah, I agree this is not a deploy blocker. Sure strange that this would all of a sudden start happening. I did change how the signout stuff works a little bit but it should mean that the redirect to sign-in page doesn't happen until the Onyx change is committed.

I just reproduced this, and the console logs around the error are sure interesting:

![image](https://user-images.githubusercontent.com/1228807/213022173-a3e00ee0-7cd7-4f3a-9d63-a44e7be18114.png)

I'm not sure what's up with that second error, maybe it's reported in another issue. Seems totally unrelated. 

The biggest oddity to me is the log `[info] Navigating to route - {"path":"/"}` happening twice. 

The pusher error is happening from `ReportActionsView` calling `Report.unsubscribeFromReportChannel(this.props.report.reportID);` when the component unmounts. So, it's almost like the component is unmounted twice, and therefore we are disconnecting from the report channel twice (one succeeds, the second one fails with that error).

**Comment by marcaaron** [comment]
Nice debugging. 

I don't have much to add but wanted to say that I am not sure if I would 100% trust that log. That can happen anytime the navigation state changes:

https://github.com/Expensify/App/blob/6902b945a5e5f2fcf0ac618e538b3b21b22c7e78/src/libs/Navigation/NavigationRoot.js#L60-L68

Doesn't answer the question of _why_ the `NavigationContainer` state changes twice - but it _might_ not be because we explicitly navigated to `/`.

**Comment by marcaaron** [comment]
Oh I wonder if the `Onyx.clear()` has something to do with it. Perhaps something about the order in which the subscribers are notified? Maybe we didn't notice this before because of this change:

https://github.com/Expensify/App/commit/6def1b37c89cdf3fd93e52cc4b68ae0984eae324

I'm assuming the `Onyx.clear()` call unhooks subscribers in a particular order - but in the previous code we would always first log the user out and _then_ clear the rest of the keys. So maybe some of those keys clearing causes a re-render somewhere. Just kinda riffing on that train of thought though - not sure how or why that would cause either of these errors.

**Comment by dylanexpensify** [comment]
Sorry I missed this -reviewing

**Comment by dylanexpensify** [comment]
Could reproduce the error in staging. @marcaaron @tgolen are we good making this external? Or do we think internal?

**Comment by tgolen** [comment]
I think it should go `external` first 👍

**Comment by dylanexpensify** [comment]
Perfect!

**Comment by esh-g** [comment]
## Proposal

**Problem**

The `socket` is being unsubscribed from more than once.

We have no `eventName` for `signOut` which causes the else condition in the following function to execute

<img width="1210" alt="Screenshot 2023-01-23 at 8 49 40 PM" src="https://user-images.githubusercontent.com/77237602/214077509-495dd649-d716-4af3-871d-c66cea627fa4.png">

And since we are unbinding from the channel, we don't need to `unsubscribe` the socket since that is happening automatically and thus this eliminates the double unsubscribing.


**Solution**

```diff
diff --git a/src/libs/Pusher/EventType.js b/src/libs/Pusher/EventType.js
index 79e7fb0c5..c2d4d7166 100644
--- a/src/libs/Pusher/EventType.js
+++ b/src/libs/Pusher/EventType.js
@@ -8,4 +8,5 @@ export default {
     EXPENSIFY_CARD_UPDATE: 'expensifyCardUpdate',
     SCREEN_SHARE_REQUEST: 'screenshareRequest',
     ONYX_API_UPDATE: 'onyxApiUpdate',
+    SIGN_OUT: 'signOut',
 };
diff --git a/src/libs/actions/Report.js b/src/libs/actions/Report.js
index 78faeaa3e..f4054d6e6 100644
--- a/src/libs/actions/Report.js
+++ b/src/libs/actions/Report.js
@@ -160,7 +160,7 @@ function unsubscribeFromReportChannel(reportID) {
 
     const pusherChannelName = getReportChannelName(reportID);
     Onyx.set(`${ONYXKEYS.COLLECTION.REPORT_USER_IS_TYPING}${reportID}`, {});
-    Pusher.unsubscribe(pusherChannelName);
+    Pusher.unsubscribe(pusherChannelName, Pusher.TYPE.SIGN_OUT);
 }
 
 const defaultNewActionSubscriber = {

```

**After fix:**

https://user-images.githubusercontent.com/77237602/214079767-e4e5850c-c512-4784-956d-125989d7f0fc.mov

Notice that in the video, there is an `[info]` log where the disconnection does happen just without the error

**Comment by s77rt** [comment]
@esh-g I don't think you are fixing the issue that way, your code won't call `socket.unsubscribe`

**Comment by esh-g** [comment]
@s77rt Look at the logs in the video, it is being unsubscribed and disconnected from just without an error

**Comment by esh-g** [comment]
Here I have console logged the socket after the unbind and you can see that it is disconnected

<img width="1257" alt="Screenshot 2023-01-23 at 9 32 56 PM" src="https://user-images.githubusercontent.com/77237602/214087704-b0b41243-3373-4c82-bce2-28609c38aaa0.png">


https://user-images.githubusercontent.com/77237602/214088123-852a2812-f161-47c8-b625-3b30198d4d51.mov

**Comment by tgolen** [comment]
@esh-g Thanks for looking into that! I have a couple of questions about your proposal.

1. If you're adding a new event type, shouldn't there also be a place where the new event type is used with `Pusher.subscribe()`?
2. Rather than creating a new event type, maybe we should just bypass calling `socket.unsubscribe(eventName)` when there is no event name? The reason I say that is because with the `eventName=''` argument defaulting to an empty string, it seems to me like someone thought it would be a valid case to not include the `eventName` but maybe they didn't run into and see this disconnect error (or they did and they felt like it wasn't an issue at the time).

**Comment by esh-g** [comment]
Yes @tgolen at first I thought the same of just removing the `socket.unsubscribe()` somehow, and as you mentioned in point 2, we should bypass calling the `socket.unsubscribe()` and I believe this proposal keeps the intention of the script intact while also bypassing the unsub because it might be a valid case where we don't want `eventName` defined this is the exact reason why didn't make any changes to the `Pusher.js` file whatsoever. 

And as for having no subscribers for the event, it is because I used it as a way to bypass the execution of `socket.unsubscribe()` because it achieved the intended effect with the least amount of changes to the original intent of the function. And maybe in the future, it can be useful to have an additional event to listen to as well!

But if you suggest for having no event for sign-out, we can achieve that as well maybe with an additional parameter to disconnect the socket. Let me know what you think!

**Comment by tgolen** [comment]
Thanks for those thoughts!

> I used it as a way to bypass the execution of socket.unsubscribe()

I think this is a sign that the solution is a workaround (aka "hack") which I think we should avoid.

> maybe in the future, it can be useful to have an additional event to listen to as well!

This is something else that should be avoided because it's a preoptimization.

I think we should just make the change to `Pusher.js`. The method is only used in two places, so it will be easy to test that the functionality works and doesn't cause regressions.

**Comment by esh-g** [comment]
Okay, then will that involve removing the `socket.unsubscribe()` altogether?

**Comment by tgolen** [comment]
I think it needs to be called when there is an `eventName`? I would just wrap it in an `if` statement.

**Comment by esh-g** [comment]
> 2. Rather than creating a new event type, maybe we should just bypass calling `socket.unsubscribe(eventName)` when 

There may be a misunderstanding between `channelName` and `eventName`. According to my understanding, `channelName` is always defined it is `socket.unsubscribe(channelName)` so could you please clarify how it would need to be called if `eventName` is defined, because right now there is an `if` to check the `eventName` and this lies in the else part?

**Comment by tgolen** [comment]
Oh, I see I have indeed confused them both 🤦 . I'm sorry about that!

I agree with you now about just removing `socket.unsubscribe()` altogether.

**Comment by esh-g** [comment]
@tgolen No problem, I have been testing the app on multiple devices by removing the `socket.unsubscribe()` and it seems to have been working as expected with no regressions. I see that line was added from about 2 years ago in this PR: https://github.com/Expensify/App/pull/35 So, I think it is safe to remove

**Comment by abdulrahuman5196** [comment]
I don't think that unsubscribeFromReportChannel() is called twice during signout as well. I tried putting a log line in the function, it was called only once during signout.

>And since we are unbinding from the channel, we don't need to unsubscribe the socket since that is happening automatically and thus this eliminates the double unsubscribing.

And this statement also might not be correct because. If so we should see the issue everytime we unsubscribe through "unsubscribeFromReportChannel" method ((called every time we take report view from screen)), but this only happens at logout.

I am trying to gather a proposal myself. Will try to bring it sooner.

**Comment by tgolen** [comment]
Wow, great analysis @Prince-Mendiratta! I love the idea of the upstream fix. I like the idea of doing that as well as solution 3 (which feels like it is using the pusher code in a better manner). @chiragsalian what do you think?

**Comment by Santhosh-Sellavel** [comment]
[Solution 3](https://github.com/Expensify/App/issues/14350#issuecomment-1400975939) solves the problem. I think that should do! 

C+ Reviewed
👀 🎀 👀

**Comment by Santhosh-Sellavel** [comment]
@Prince-Mendiratta Please always wait until the Job is assigned to you or go ahead for PR from team. 

I'll wait for @chiragsalian thoughts too, thanks!

**Comment by chiragsalian** [comment]
Neat, I'm just catching up here, and yeah great work on the analysis @Prince-Mendiratta.
Yeah like tim suggested, I really like the idea of doing your upstream fix and solution 3.

Proposals LGTM. 

> Should I close the PR and wait for @chiragsalian to share his thoughts first?

Nope, you're good. Next time remember to wait for C+ and Engineer to approve. This is only because if either disagrees on the proposal you wouldn't waste time working on PRs.

**Comment by Santhosh-Sellavel** [comment]
@chiragsalian I'm not sure what caused this regression, apart from the PR which removed it https://github.com/Expensify/App/pull/752 did we break something else?

cc: @Prince-Mendiratta

**Comment by chiragsalian** [comment]
Hmm okay, so I think your solution was fine @Prince-Mendiratta.

You mention your fix might have brought back this issue - https://github.com/Expensify/App/pull/752. But i was not able to reproduce it. The issue is more than 2 years old.

With your solution prince I would think that unbinding from the event won't unsubscribe from the channel itself and maybe that's fine. We can create a follow-up issue to figure out how to unsubscribe from the channel itself so that it doesn't take up pusher space perhaps.

> has this issue been in the repo since 3 years or did it just start popping up?!

Not sure, maybe no one noticed it before. Hard to say 🤷‍♂️  Not sure if it's worth the time and effort tracing through the past code IMO.

As a first step let's confirm if there are problems,
1. Tagging @marcaaron if you can reproduce this [issue](https://github.com/Expensify/App/pull/752) on the latest App code since you created the issue to begin with. I could not reproduce the issue.
2. To confirm if there is a problem only unsubscribing from the event but keeping the socket active. If there is no problem here I'm not sure if anything needs to be done.

**Comment by Santhosh-Sellavel** [comment]
Cool, @chiragsalian We can this off checklist!

**Comment by chiragsalian** [comment]
Yup, updated

**Comment by marcaaron** [comment]
>Tagging @marcaaron if you can reproduce this https://github.com/Expensify/App/pull/752 on the latest App code since you created the issue to begin with. I could not reproduce the issue.

Sorry, I won't have time to test this. If we can't reproduce it then let's consider it fixed.

**Comment by Santhosh-Sellavel** [comment]
@dylanexpensify bump for payment thanks!

**Comment by dylanexpensify** [comment]
on it!

**Comment by dylanexpensify** [comment]
@Santhosh-Sellavel did you apply? Can't see you on the job proposals in Upwork

**Comment by dylanexpensify** [comment]
@Prince-Mendiratta sent offer!

**Comment by dylanexpensify** [comment]
@Prince-Mendiratta can you confirm your PR didn't lead to a regression? If so, then yes it's eligible!

**Comment by dylanexpensify** [comment]
Wooot!

**Comment by dylanexpensify** [comment]
payment sent @Prince-Mendiratta with bonus!
offer sent @Santhosh-Sellavel !

**Comment by dylanexpensify** [comment]
payment sent @Santhosh-Sellavel !

**Comment by dylanexpensify** [comment]
posted about RT steps

**Comment by dylanexpensify** [comment]
alrighty, we're gonna do nothing for RT, so all good here!
